// react-dims/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withContext = exports.Provider = void 0;

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj["default"] = obj; return newObj; } }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var ReactDims = _react["default"].createContext(null);

var Provider = function Provider(props) {
  var domNode = (0, _react.useRef)(null);

  var _useState = (0, _react.useState)({}),
      _useState2 = _slicedToArray(_useState, 2),
      dimensions = _useState2[0],
      setDimensions = _useState2[1];

  var _useState3 = (0, _react.useState)(0),
      _useState4 = _slicedToArray(_useState3, 2),
      invocationContext = _useState4[0],
      setInvocationContext = _useState4[1];

  (0, _react.useEffect)(function () {
    setDimensions(domNode.current.getBoundingClientRect());
  }, []);
  (0, _react.useEffect)(function () {
    window.addEventListener('resize', getNodeDimensions);
    return function () {
      window.removeEventListener('resize', getNodeDimensions);
    };
  }, []);

  var getNodeDimensions = function getNodeDimensions() {
    clearTimeout(invocationContext);
    setInvocationContext(function () {
      return setTimeout(function () {
        setDimensions(domNode.current.getBoundingClientRect());
      }, props.debounce);
    });
  };

  return _react["default"].createElement("div", {
    ref: domNode,
    style: {
      height: '100%'
    }
  }, _react["default"].createElement(ReactDims.Provider, {
    value: dimensions
  }, props.children));
};

exports.Provider = Provider;
Provider.propTypes = {
  debounce: _propTypes["default"].number
};
Provider.defaultProps = {
  debounce: 100
};

var withContext = function withContext(ChildComponent) {
  return function (props) {
    return _react["default"].createElement(ReactDims.Consumer, null, function (incomingDims) {
      return _react["default"].createElement(ChildComponent, _extends({}, props, {
        dims: incomingDims
      }));
    });
  };
};

exports.withContext = withContext;
